### ISSUE TITLE
[Feature request]: Native ARM32/ARM64 Support for Windows (or universal app)

### ISSUE THREAD

**Issue by holaimscott** [project64/project64#2217]
### Request box

It would mean a lot if Project64 can continue to be the best emulator for Nintendo 64 by adding native ARM64 build on Windows, or consider building their app to run across all windows platforms (x86,ARM)

### Reason box

With the rise of ARM devices coming up, having native ARM64 support can help Project64 be ahead of the curve and bring further compatibility with future systems, as well as being able to run in current ARM devices natively.

A lot of the video plugins used in Project64 work in ARM devices (Android), and to my understanding Project64 already has an ARM compiler sine there is an android app version too.

### Examples

Several ARM devices are powerful enough to emulate Nintendo 64 now, however when using x86 emulators, performance can drop drastically.

Microsoft is not the only one with fully working ARM devices, since now there are a lot of community projects allowing devices that are powerful enough to run Windows [and show their true potential.](https://youtu.be/6NADjIc9uNk)

A lot of popular and long-time emulators [like Dolphin, Duckstation and PPSSPP](https://youtu.be/93grsZl3lHo) have native ARM support. Having Project64 optimized and/or be built for ARM64 devices would be splendid.

**Comment by DerekTurtleRoe** [comment]
@holaimscott Currently Project64 only builds on Windows. Zilmar has plans for an ARM64 version of Project64, but there is no set time/release date for that. Once it's ported to Android, supporting other operating systems like Mac and Linux should be fairly straight forward, but no time line on that either. Making all the code portable and removing Windows-specific code and such will take quite a bit of time, since I think the emulator was pretty much designed with Windows in mind.

Not to mention Project64 relies on plugins, so those have to be ported as well.

**Comment by flagrama** [comment]
Windows on ARM exists though Derek. Since they are talking about x86 emulators dropping performance it sounds like they're directly talking about it.

**Comment by DerekTurtleRoe** [comment]
@flagrama That would still rely on porting everything to ARM wouldn't it though?

**Comment by flagrama** [comment]
Yes, but your comment seems to ignore it exists and talks about other OSes when it seems like the original request was for a version of Windows.

**Comment by DerekTurtleRoe** [comment]
> Yes, but your comment seems to ignore it exists and talks about other OSes when it seems like the original request was for a version of Windows.

I guess my point was that when the Android port is going on, it could be added then maybe, but I'm not as familiar with Windows on ARM as I am with Android and other OS's.

**Comment by project64** [comment]
It exists on arm .. but that has problems and needs more work, I am at the moment working on moving memory handling in to handles to remove use of exceptions which should make it easier to build recompilers on other platforms. 

There is x64, arm and arm64 to fix up.

Once all that is done then I can look at arm machine, not sure if I would have to buy one to confirm it works correctly, though my phone might be enough

**Comment by holaimscott** [comment]
> It exists on arm .. but that has problems and needs more work, I am at the moment working on moving memory handling in to handles to remove use of exceptions which should make it easier to build recompilers on other platforms.
> 
> There is x64, arm and arm64 to fix up.
> 
> Once all that is done then I can look at arm machine, not sure if I would have to buy one to confirm it works correctly, though my phone might be enough

This is amazing news, thank you so much!  I am excited to see yall are working on those three platforms in Windows (x64,ARM,ARM64). Hope everything needed turns out well

To my understanding, Qualcomm and Microsoft also [have a developer kit for ARM.](https://www.microsoft.com/en-us/d/ecs-liva-mini-box-qc710-desktop/8z247h1h3skp?activetab=pivot:overviewtab) but any arm device should work.

If your team also needs ARM testers in the future, I am a part of a community that has people that have a handheld [running Windows ARM on a Snapdragon 845 gaming handheld.](https://www.reddit.com/r/OdinHandheld/comments/tf3m74/win_odin_quick_facts/) that would love to help out, myself included.

Again, thank you :)

**Comment by dftf-stu** [comment]
_Project64_ doesn't even have a stable 64-bit (AMD64 / x86-64) build yet... and given the vast, vast majority of all _Windows_ installs nowadays are that architecture, I'd say focus on getting that working first before considering an ARM64 build.

As for the ARM32 build, I'd argue that's not worth bothering with, given:
- _Windows 11_ (and all future versions of _Windows_) now only support 64-bit AMD, ARM and Intel CPUs
- _Windows RT_ (aka "_Windows 8 on ARM_") ends-support in January 2023; and _Windows 10 on ARM_ in October 2025.  After that last date, Microsoft will have no 32-bit ARM OSes still in-support
- Apple's M1 and M2 chips ("Apple Silicon") are 64-bit only ARM CPUs, and macOS 10.15 ("Catalina"), released October 2019, drops support for all 32-bit code
- Current versions of iOS and iPadOS are ARM64 only; Android will also soon only be available in ARM64 builds

**Comment by DerekTurtleRoe** [comment]
Yeah, getting ARM support well enough for Android is something Zilmar is working on I think. But the priority should be (approximately):

32-bit/64-bit Windows
Android
Linux

IDK if we ever wanna go down the Apple route, that's another beast. Not sure how other FOSS handles that but I know it isn't easy.
